File: example-2.5/scratch.py
# library(dplyr)
# library(ggplot2)
# library(magrittr)
# library(purrr)
# library(ISLR)

import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt

df = pd.read_csv("data-auto.csv")
df.origin = df.origin.map({1: 'America', 2: 'Europe', 3: 'Japan'})

# Generate a plot to understand how things look.
pd.plotting.scatter_matrix(df.iloc[:, 0:8], alpha = 0.4)
plt.show()

col_names = df.columns.to_list()
form_1 = 'mpg ~ ' + ' + '.join(col_names[1:8])
# May want to address the multicollinearity here but not sure.
# form_1 = form_1.replace(' + displacement', '').replace(' + horsepower', '')

# TODO Need to generate the diagnostic plots for the model fit but it is
# unclear how much of this we can do without them having seen it in lectures

fit_1 = smf.ols(formula = form_1, data = df).fit()

form_2 = form_1.replace('weight', 'np.log(weight)').replace('acceleration', 'np.power(acceleration, 2)')

fit_2 = smf.ols(formula = form_2, data = df).fit()

form_3 = form_2 + ' + origin * year'

fit_3 = smf.ols(formula = form_3, data = df).fit()

File: example-2.5/scratch.R
library(dplyr)
library(ggplot2)
library(magrittr)
library(purrr)
library(ISLR)

#' Look at the data and ensure that the inferred types make sense.

Auto |> select(-origin, -name, -cylinders) |> pairs()
Auto |> select(-origin, -name, -cylinders) |> cor()
str(Auto)
df1 <- Auto |> mutate(origin = as.factor(origin))

#' Fit a boring model and look at what you get.
m1 <- lm(mpg ~ . - name, data = df1)
summary(m1)

plot(m1)

#' Transform some of the variables to make their relationship with MPGs look
#' more linear.
df2 <- df1 |>
  mutate(ln_weight = log(weight),
         sqrd_acceleration = acceleration ^ 2)

m2 <- lm(mpg ~ . - name - weight - sqrd_acceleration, data = df2)
summary(m2)

#' Observe that removing horsepower leaves a very similar model probably because
#' horsepower is so closely associated with the displacement.
m3 <- lm(mpg ~ . - name - weight - acceleration, data = df2)
summary(m3)

#' Observe that you can also do this with the original data set by modifying the formula.
m3b <- lm(mpg ~ . - name - weight + log(weight) - acceleration + I(acceleration^2), data = df1)
summary(m3b)

## Observe that if you consider the interaction between year and the country
## that you find that the MPG is improving faster in Europe and Japan than in
## America.
m4a <- lm(mpg ~ . - name - weight + log(weight) - acceleration + I(acceleration^2), data = df1)
summary(m4a)

m4b <- lm(mpg ~ . + origin*year - name - weight + log(weight) - acceleration + I(acceleration^2), data = df1)
summary(m4b)
